main.go

```diff
@@ -951,3 +903,252 @@ func addControllers(
+
+				constraintsReconciler, constraintEvents := depclient.NewControllerRuntimeSource()
+
+				constraintsWatcher, err := depclient.New(managedCfg, constraintsReconciler, nil)
+				if err != nil {
+					log.Error(err, "Unable to create the constraints watcher. Will retry.")
+
+					break
+				}
+
+				// Create a separate context for the dynamic watcher to be able to cancel that separately in case
+				// starting the manager needs to be retried.
+				dynamicWatcherCtx, dynamicWatchCancel := context.WithCancel(ctx)
+				defer dynamicWatchCancel()
+
+				go func() {
+					err := constraintsWatcher.Start(dynamicWatcherCtx)
+					if err != nil {
+						panic(err)
+					}
+				}()
+
+				// Wait until the constraints watcher has started.
+				<-constraintsWatcher.Started()
+
+				if err = (&gatekeepersync.GatekeeperConstraintReconciler{
+					Client: mgr.GetClient(),
+					ComplianceEventSender: utils.ComplianceEventSender{
+						ClusterNamespace: tool.Options.ClusterNamespace,
+						ClientSet:        clientset,
+						ControllerName:   gatekeepersync.ControllerName,
+						InstanceName:     instanceName,
+					},
+					DynamicClient:        dynamicClient,
+					ConstraintsWatcher:   constraintsWatcher,
+					Scheme:               mgr.GetScheme(),
+					ConcurrentReconciles: int(tool.Options.EvaluationConcurrency),
+				}).SetupWithManager(mgr, constraintEvents); err != nil {
+					log.Error(
+						err, "Unable to create controller. Will retry.", "controller", gatekeepersync.ControllerName,
+					)
+
+					// Stop the dynamic watcher since the manager will get recreated.
+					dynamicWatchCancel()
+
+					break
+				}
+
+				// Add the health bind address to be considered by the health proxy.
+				healthAddressesLock.Lock()
+				healthAddresses[healthAddress] = true
+				healthAddressesLock.Unlock()
+
+				log.Info("Gatekeeper is installed. Starting the gatekeeper-constraint-status-sync controller.")
+
+				// This blocks until the manager stops.
+				err = mgr.Start(ctx)
+
+				// Remove the health bind address after the manager stops.
+				healthAddressesLock.Lock()
+				delete(healthAddresses, healthAddress)
+				healthAddressesLock.Unlock()
+
+				if err != nil {
+					log.Error(err, "Unable to start the Gatekeeper constraint status sync manager. Will retry.")
+
+					// Stop the dynamic watcher since the manager will get recreated.
+					dynamicWatchCancel()
+
+					break
+				}
+
+				return
+			}
+
+			time.Sleep(time.Second)
+		}
+	}
+
+	w, gatekeeperInstalled, err := newRetryWatcher()
+	if err != nil {
+		return err
+	}
+
+	var mgrCtx context.Context
+	var mgrCtxCancel context.CancelFunc
+
+	mgrRunning := false
+
+	if gatekeeperInstalled {
+		mgrRunning = true
+
+		mgrCtx, mgrCtxCancel = context.WithCancel(ctx)
+
+		go startMgr(mgrCtx)
+	}
+
+	go func() {
+		for {
+			select {
+			case <-w.Done():
+				select {
+				case <-ctx.Done():
+					// Stop the retry watcher if the parent context is canceled.
+					w.Stop()
```

Comment:
Since this is inside the `<- w.Done()` case, isn't the watcher already stopped?

Reply:
I'm not certain about this. I couldn't find any handling of a closed context in the client-go code but it's not harmful to include this since it explicitly tells the retry loop to stop.